### PR TITLE
Feature/122 add ability to use layered concentric infinite cylinders with refractive index mismatch

### PR DESCRIPTION
**Pull request by hayakawa16**
Hi, this is a *draft* of a PR so I can see file differences and make sure all looks good.  I will assign reviewers when I'm ready for review.

### CODE REVIEW

**Comment by hayakawa16** [comment]
I'm ready for review.

**Comment by hayakawa16** [comment]
I just saw sonarcloud review.  Give me a bit to address these.

**Comment by hayakawa16** [comment]
Okay I'm ready for review.

**Review by janakarana — APPROVED** [review]
Everything looks fine to me

File: src/Vts.Test/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidationTests.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                    }
+                ),
+                new List<IDetectorInput>() 
+                {
+                    new FluenceOfXAndYAndZDetectorInput()
+                }
+            );
+            var result = SimulationInputValidation.ValidateInput(input);
+            Assert.IsFalse(result.IsValid);
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidationTests.cs:51]: This is nitpicky but I like to see a space between methods and therefore a line feed after the method and before the comments for the next method. I use space to help me find the individual methods.

File: src/Vts.Test/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidationTests.cs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                    }
+                ),
+                new List<IDetectorInput>()
+                {
+                    new FluenceOfXAndYAndZDetectorInput()
+                }
+            );
+            var result = SimulationInputValidation.ValidateInput(input);
+            Assert.IsFalse(result.IsValid);
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidationTests.cs:89]: Line feed

File: src/Vts.Test/MonteCarlo/Tissues/MultiConcentricInclusionTissueTests.cs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            photon = new Photon( // on bottom of slab pointed out
+                new Position(0, 0, 100.0),
+                new Direction(0.0, 0, 1.0),
+                1.0,
+                _tissue,
+                1,
+                new Random());
+            index = _tissue.GetNeighborRegionIndex(photon);
+            Assert.AreEqual(2, index);
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/Tissues/MultiConcentricInclusionTissueTests.cs:116]: Line feed

File: src/Vts.Test/MonteCarlo/Tissues/MultiConcentricInclusionTissueTests.cs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                new Position(0, 0, 1.0),
+                new Direction(0.0, 0, 1.0),
+                1,
+                _tissue,
+                1,
+                new Random());
+            var cosTheta = _tissue.GetAngleRelativeToBoundaryNormal(photon);
+            Assert.AreEqual(1,cosTheta);
+        }
+
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/Tissues/MultiConcentricInclusionTissueTests.cs:284]: Extra line feed

File: src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        /// </summary>
+        [OneTimeSetUp]
+        [OneTimeTearDown]
+        public void Clear_folders_and_files()
+        {
+            foreach (var file in _listOfTestGeneratedFiles)
+            {
+                FileIO.FileDelete(file);
+            }
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs:32]: Line feed

File: src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                        new LayerTissueRegion(
+                            new DoubleRange(100.0, double.PositiveInfinity),
+                            new OpticalProperties(0.0, 1e-10, 1.0, 1.0))
+                    }
+                );
+
+            var iCloned = i.Clone();
+
+            Assert.AreEqual(iCloned.Regions[1].RegionOP.Mua, i.Regions[1].RegionOP.Mua);
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs:75]: Line feed

File: src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                        new LayerTissueRegion(
+                            new DoubleRange(100.0, double.PositiveInfinity),
+                            new OpticalProperties(0.0, 1e-10, 1.0, 1.0))
+                    }
+                );
+            i.WriteToJson("SingleInfiniteCylinderTissue.txt");
+            var iCloned = FileIO.ReadFromJson<SingleInfiniteCylinderTissueInput>("SingleInfiniteCylinderTissue.txt");
+
+            Assert.AreEqual(iCloned.Regions[1].RegionOP.Mua, i.Regions[1].RegionOP.Mua);
+        }
```
Comment (lmalenfant) [inline src/Vts.Test/MonteCarlo/Tissues/SingleInfiniteCylinderTissueTests.cs:104]: Line feed

File: src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            // check that there is at least one layer of tissue 
+            var layerTissueRegions = tissueLayers as LayerTissueRegion[] ?? tissueLayers.ToArray();
+            if (!layerTissueRegions.Any())
+            {
+                tempResult = new ValidationResult(
+                    false,
+                    "SingleInfiniteCylinderTissueInput: tissue layer is assumed to be at least a single layer with air layer above and below",
+                    "SingleInfiniteCylinderTissueInput: redefine tissue definition to contain at least a single layer of tissue");
+            }
+            if (!tempResult.IsValid) { return tempResult; }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs:62]: We don't need the braces here if it is on one line.

File: src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (!correctlyContainedInLayer)
+            {
+                tempResult = new ValidationResult(
+                    false,
+                    "SingleInfiniteCylinderTissueInput: infinite cylinders must be entirely contained within a tissue layer",
+                    "Resize radii and or Center so that infinite cylinders entirely within tissue layer");
+            }
+
+            if (!tempResult.IsValid) { return tempResult; }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs:81]: We don't need the braces if the code is on one line.

File: src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        /// of code.
+        /// </summary>
+        /// <param name="layers">list of LayerTissueRegion</param>
+        /// <param name="infiniteCylinder">InfiniteCylinderTissueRegion</param>
+        /// <returns>An instance of the ValidationResult class</returns>
+        private static ValidationResult ValidateGeometry(IList<LayerTissueRegion> layers, InfiniteCylinderTissueRegion infiniteCylinder)
+        {            
+            // check that layer definition is valid
+            var tempResult = MultiLayerTissueInputValidation.ValidateLayers(layers);
+            if (!tempResult.IsValid){ return tempResult; }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs:39]: We don't need the braces if the code is on one line.

File: src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (infiniteCylinder.Radius == 0.0)
+            {
+                tempResult = new ValidationResult(
+                    false,
+                    "SingleInfiniteCylinderTissueInput: infinite cylinder has radius equal to 0",
+                    "SingleInfiniteCylinderTissueInput: make sure infinite cylinder radius is > 0");
+            }
+            if (!tempResult.IsValid) { return tempResult; }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/DataStructuresValidation/TissueInputs/SingleInfiniteCylinderTissueInputValidation.cs:48]: We don't need the braces if the code is on one line.

**Review by lmalenfant — COMMENTED** [review]
Most of these are just suggestions so I have no problem approving if you don't want to fix them now.

**Comment by hayakawa16** [comment]
Hi @lmalenfant, thanks for your review. I can make these edits.

**Comment by hayakawa16** [comment]
I have updated code and pushed.

On another topic.  I pulled master and ran an infile that I ran in August 2023 and obtained identical results.  Then I pulled this branch and ran the same infile.  The results are not identical. I think they are statistically equivalent but not sure.  I'd like to know for sure before I merge this branch into master.

**Comment by lmalenfant** [comment]
I also noticed that we cannot merge right now because I made the master branch read only. We either have to force push or remove the lock. I checked that box so we can't push to master without a PR. I need to do more research to figure out the best protection.

**Comment by hayakawa16** [comment]
I've performed my analysis of results using master and this branch and they agree within variance.

**Comment by hayakawa16** [comment]
I feel okay about merging this into master.  Should I check "Merge without waiting for requirements to be met (bypass branch protections)" and click "Merge pull request"?

**Review by lmalenfant — APPROVED** [review]

**Comment by lmalenfant** [comment]
@hayakawa16 yes go ahead and merge, I think you just need to check the box and then the button will be available.

**Pull request merged**